tests/testsuite/git.rs

```diff
@@ -1828,6 +1845,729 @@ fn fetch_downloads() {
+) -> anyhow::Result<()> {
+    // Example where an old lockfile with an explicit branch="master" in Cargo.toml.
+    Package::new("bar", "1.0.0").publish();
+    let (bar, bar_repo) = git::new_repo("bar", |p| {
+        p.file("Cargo.toml", &basic_manifest("bar", "1.0.0"))
+            .file("src/lib.rs", "")
+    });
+
+    bar.change_file("src/lib.rs", "// change");
+    git::add(&bar_repo);
+    git::commit(&bar_repo);
+
+    {
+        let mut walk = bar_repo.revwalk()?;
+        walk.push_head()?;
+        assert_eq!(
+            walk.count(),
+            2,
+            "original repo has initial commit and change commit"
+        );
+    }
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            &format!(
+                r#"
+                    [package]
+                    name = "foo"
+                    version = "0.1.0"
+
+                    [dependencies]
+                    bar = {{ version = "1.0", git = "{}", branch = "master" }}
+                "#,
+                bar.url()
+            ),
+        )
+        .file("src/lib.rs", "")
+        .build();
+
+    p.cargo("update")
+        .arg("-Zgitoxide=fetch,shallow-deps")
+        .masquerade_as_nightly_cargo(&["unstable features must be available for -Z gitoxide"])
+        .run();
+
+    let db_clone = gix::open_opts(
+        glob::glob(paths::home().join(".cargo/git/db/bar-*").to_str().unwrap())?
+            .next()
+            .unwrap()?,
+        gix::open::Options::isolated(),
+    )?;
+    assert!(db_clone.is_shallow());
+    assert_eq!(
+        db_clone
+            .rev_parse_single("origin/master")?
+            .ancestors()
+            .all()?
+            .count(),
+        1,
+        "db clones are shallow and have a shortened history"
+    );
+
+    let dep_checkout = gix::open_opts(
+        glob::glob(
+            paths::home()
+                .join(".cargo/git/checkouts/bar-*/*/.git")
+                .to_str()
+                .unwrap(),
+        )?
+        .next()
+        .unwrap()?,
+        gix::open::Options::isolated(),
+    )?;
+    assert!(dep_checkout.is_shallow());
+    assert_eq!(
+        dep_checkout.head_id()?.ancestors().all()?.count(),
+        1,
+        "db checkouts are hard-linked clones with the shallow file copied separately."
+    );
+
+    bar.change_file("src/lib.rs", "// another change");
+    git::add(&bar_repo);
+    git::commit(&bar_repo);
+    {
+        let mut walk = bar_repo.revwalk()?;
+        walk.push_head()?;
+        assert_eq!(
+            walk.count(),
+            3,
+            "original repo has initial commit and change commit, and another change"
+        );
+    }
+
+    p.cargo("update")
+        .env("__CARGO_USE_GITOXIDE_INSTEAD_OF_GIT2", "0")
+        .run();
+
+    let db_clone = gix::open_opts(
+        glob::glob(paths::home().join(".cargo/git/db/bar-*").to_str().unwrap())?
```

Comment:
That's another great catch, because these glob statements implicitly rely on the sort order that `glob` kind of implicitly performs, which result in non-shallow clones to be returned first. This happens to be something the code leverages.

It took me a while to figure out what to do here and decided to factor these 'single-path' lookups out into more descriptive functions. One of these now supports filtering as well and the call sites look much more descriptive.

This is implemented in 449116558d12f39c5f921ec5d64ea9d170da61bc.